src/sc/sidechain.h

```diff
@@ -55,49 +57,65 @@ class ScInfo
         READWRITE(creationTxHash);
         READWRITE(balance);
         READWRITE(creationData);
-//        READWRITE(sBackwardTransfers);
+        READWRITE(vImmatureAmounts);
     }
 };
 
 
 typedef boost::unordered_map<uint256, ScInfo, ObjectHasher> ScInfoMap;
 
+class ScCoinsViewCache
+{
+    ScInfoMap mUpdate;
+    std::set<uint256> sErase;
+    std::set<uint256> sDirty;
+
+public:
+    bool UpdateScCoins(const CTransaction& tx, const CBlock&, int nHeight, CTxUndo& txundo);
+    bool UndoScCreation(const CTxUndo& undo);
+    bool IncrementScBalance(int nHeight, CBlockUndo& blockundo);
+    bool DecrementScBalance(int nHeight, CBlockUndo& blockundo);
+
+    const ScInfoMap& getUpdateMap() const { return mUpdate; }
+    bool sidechainExists(const uint256& scId) const;
+
+    bool Flush();
+
+    ScCoinsViewCache();
+    ScCoinsViewCache(const ScCoinsViewCache&) = delete;
+    ScCoinsViewCache& operator=(const ScCoinsViewCache &) = delete;
+    ScCoinsViewCache(ScCoinsViewCache&) = delete;
+    ScCoinsViewCache& operator=(ScCoinsViewCache &) = delete;
+};
+
 class ScMgr
 {
   private:
     // Disallow instantiation outside of the class.
-    ScMgr(): db(NULL), bVerifyingDb(false) {}
+    ScMgr(): db(NULL) {}
     ~ScMgr() { delete db; }
 
-    CCriticalSection sc_lock;
+    mutable CCriticalSection sc_lock;
     ScInfoMap mScInfo;
     CLevelDBWrapper* db;
-    bool bVerifyingDb;
 
     // low level api for DB
+    friend class ScCoinsViewCache;
     bool writeToDb(const uint256& scId, const ScInfo& info);
     void eraseFromDb(const uint256& scId);
 
-    // add/remove/find obj in sc map and DB
-    bool addSidechain(const uint256& scId, const ScInfo& info);
-    void removeSidechain(const uint256& scId);
-    bool addScBackwardTx(const uint256& scId, const uint256& hash);
-    bool removeScBackwardTx(const uint256& scId, const uint256& hash);
-    bool containsScBackwardTx(const uint256& scId, const uint256& txHash);
-
     bool checkSidechainCreation(const CTransaction& tx, CValidationState& state);
     bool hasSCCreationConflictsInMempool(const CTxMemPool& pool, const CTransaction& tx);
```

Comment:
It's present 3 different naming style for sidechains related logic:
1) `Sc`, for example `ScInfo`
2) `SC`, for example `hasSCCreationConflictsInMempool`
3) `Sidechain`, for example checkSidechainCreation
why not to use a single naming?

Reply:
Ok, using Sc more uniformly